kedro/io/kedro_data_catalog.py

```diff
@@ -231,40 +272,77 @@ def _get_dataset(
     def add(
         self, ds_name: str, dataset: AbstractDataset, replace: bool = False
     ) -> None:
+        # TODO: remove when removing old catalog
         """Adds a new ``AbstractDataset`` object to the ``KedroDataCatalog``."""
-        if ds_name in self._datasets:
-            if replace:
-                self._logger.warning("Replacing dataset '%s'", ds_name)
-            else:
-                raise DatasetAlreadyExistsError(
-                    f"Dataset '{ds_name}' has already been registered"
-                )
-        self._datasets[ds_name] = dataset
-
-    def list(self, regex_search: str | None = None) -> list[str]:
+        if ds_name in self._datasets and not replace:
+            raise DatasetAlreadyExistsError(
+                f"Dataset '{ds_name}' has already been registered"
+            )
+        self.__setitem__(ds_name, dataset)
+
+    def list(self, regex_search: str | None = None) -> List[str]:  # noqa: UP006
+        # TODO: remove when removing old catalog
         """
         List of all dataset names registered in the catalog.
         This can be filtered by providing an optional regular expression
         which will only return matching keys.
         """
-
         if regex_search is None:
-            return list(self._datasets.keys())
+            return self.keys()
 
-        if not regex_search.strip():
+        if regex_search == "":
             self._logger.warning("The empty string will not match any datasets")
             return []
 
+        return self.filter(regex_search)
+
+    def filter(
```

Comment:
Why I am usually in favour of simpler names, here `filter` is a bit ambiguous, since it's unclear what we are filtering exactly. If we are to keep a separate method, I'd rather keep `list()` with the regex.

Reply:
Done

Reply:
I quite disagree with this one, because "list" already existed in previous version and virtually no one new about it being able to filter with regex as shown in the interview, so I'd argue that's not helping discovery. 

I personally find `filter` more explicit, or even the original `.values(regex=...)`. What do other think?

Reply:
@Galileo-Galilei we agreed to rename it based on the solution for https://github.com/kedro-org/kedro/issues/3917, so we don't rename it several times. That's why we kept the old name for now.